File reader script

aeb0 opened the pull request:
Intended for generating csvs that can be easily read into postgres.
After looking at the slack export, it looked like there are three logical pieces of data that lend themselves to SQL tables.  Thus the script ouputs one csv file per table.

TODO:
1) Remove duplicates from the user before writing to file
2) Escape commas in messages and attachments since they will screw up the comma delimiting of the csv.  Or delimit with tab or something

Commit:
script to clean export slack json into csv data

Need to escape commas in messages and attachments since they will
screw up the comma delimiting of the csv.  Intended for generating
csvs that can be easily read into postgres.

Commit:
include sample data for testing and context

mLuby commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+  fs.writeFileSync(__dirname.concat('/users-'.concat(outputPathSuffix)), messagesCSV, 'utf8');
+
+  console.log(`writing attachments to csv: ${__dirname.concat('/attachments-'.concat(outputPathSuffix))}`);
+  fs.writeFileSync(__dirname.concat('/attachments-'.concat(outputPathSuffix)), messagesCSV, 'utf8');
```
No need for these `fs.writeFile`s to be sync as they don't have any after-effects.

Also do you mean to be saving the same messagesCSV to three places?

aeb0 replied:
Oops, no supposed to be messagesCSV, attachmentsCSV, usersCSV. fixed.

aeb0 replied:
Regarding the sync vs async...My thinking was sync might be safer than async, as this script may be triggered inside of a loop that iterates over the thousands of json files in the slack export.  If its sync node will block on each iteration preventing JSON.parse from being called on a ton of shit at the same time and maybe causing memory issues.  

Probably not a serious concern though. Do you have a reason for suggesting async?

mLuby replied:
That sounds like a good reason for sync.

mLuby commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+   *    |servicename|title|titlelink|text|fromurl|
+   */
+
+  parsedMessages.forEach((message) => {
```
This is a `reduce` operation, taking `message` objects and returning `messagesCSV`, `usersCSV`, and `attachmentsCSV` strings. 🤓

aeb0 replied:
good point. fixed in refactor

mLuby commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+    return;
+  }
+
+  const messages = fs.readFileSync(path, 'utf8');
```
Everything else could happen as the `async` version's callback, right?

mLuby commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+  *  One JSON file from Slack export in, 3 csv files out
+  *  We ignore parts of the data people are unlikely to care about in order
+  *  to save space: channel join messages, reactions, edit messages,
+  *  pinned items, certain attachment metadata
```
I for one feel reactions are important, but we can always add that in after MVP right?

aeb0 replied:
We could, also not a big deal to add them in right now.

mLuby commented on sample-data/2016-11-03.json:
```diff
@@ -0,0 +1,53 @@
+        "subtype": "channel_join",
+        "ts": "1478218946.000203"
+    }
+]
```
missing newline

dting commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+  parsedMessages.forEach((message) => {
+    if (message.subtype === 'channel_join' || message.subtype === 'pinned_item') return;
+
+    let messagesRow    = [];
```
`const`s here?

dting commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+    if (message.ts)        messagesRow[2] = message.ts;
+    if (message.purpose)   messagesRow[3] = true;
+    if (message.attachments) {
+      message.attachments.forEach((attachment) => {
```
Can there be multiple attachments per message? If so, does this just leave you with the last attachment's info in attachmentsRow?

aeb0 replied:
there could be good, point and also yes that was a bug. fixed in refactor :)

dting commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+   *
+   *    Attachments
+   *    |servicename|title|titlelink|text|fromurl|
+   */
```
Does this leave us with 3 separate `csv`s with no way to associate which message each attachment came from and duplicate entries in the users `csv`?

dting replied:
I totally should have read the comments... Sorry I was in a rush this morning. Ignore the dupes remark.

dting commented on scripts/json-to-csv.js:
```diff
@@ -0,0 +1,97 @@
+      attachmentsRow[5] = channel;
+      attachmentsCSV += attachmentsRow.join(',').concat('\n');
+    }
+  });
```
I feel like we should be writing these out as they are processed or the entire contents of the file needs to be read into memory before we write anything out. I haven't done any real research but a library like https://github.com/maxogden/csv-write-stream would probably work.

mLuby commented:
re removing dupes—can be done in bash pretty quickly.

``` bash
sort users-file.sql | uniq
```

A little trickier in JS.

``` js
dedup(usersCSV.split('\n')))
function dedup (listOfStrings) { return Object.keys(listOfStrings.reduce((uniques, string) => Object.assign(uniques, {[string]: true}), {})) }
```

re delimiting, also escape newlines?

Commit:
Refactored to use reduce and improve readability

Each csv file is now produced by its own function call for
legibility.
Fixed a bug causing multiple attachments on the same message to be
treated as a single attacment
Strip commas and newlines from text since they are problematic

dting commented:
Hmm. I'm kind of wondering why even go to the trouble of writing this out to csv when we can just batch insert into pg. Are we going to use these files somewhere else? If not just inserting them directly into pg seems to make more sense to retain their relationships.

aeb0 commented:
@dting can you explain more on if batch insert would work with json? And into a document store or relational?  (I am operating on the assumption we are going relational based on a couple of articles)  Googled around a bit but didn't find anything that helpful re: batch insert.

mLuby commented on scripts/json-to-csv.js:
```diff
@@ -38,58 +32,89 @@ function cleanSlackJSON(path, channel) {
-      attachmentsCSV += attachmentsRow.join(',').concat('\n');
-    }
-  });
+   let messagesCSV = parsedMessages.reduce((accum, currMessage) => {
```
Close to what I meant. I think you can avoid some duplication via destructuring:

``` js
const {messagesCSV, usersCSV, attachmentsCSV} = parsedMessages.reduce((accum, currMessage) => {
```

aeb0 commented:
Holding on merging this while we figure out if we can use [another solution](https://github.com/bsvingen/slack2pg) to handle importing more easily

mLuby commented:
Any reason not to merge this and if that other solution pans out submit a new PR replacing this with that? Iterative! Agile!
